# 01. Import libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
# Allocating variable to folder path

path = r'C:\Users\Amy Yip\Desktop\Career Foundry Data Analytics\Achievement 4 Python\05-02-2022 Instacart Basket Analysis\02 Data'

In [3]:
# Importing wrangled orders.csv data

df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled_ex4.4.csv'), index_col = False)

In [4]:
# Importing products.csv data

df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col = False)

# 03. Data Consistency Check on df_prods from 4.5 Course

## Check for Mixed-Type Data

In [5]:
# Create a dataframe called df_test

df_test = pd.DataFrame()

In [6]:
# Create a mixed type column (with numeric, string, boolean values)

df_test['mixed'] = ['a', 'b', 1, True]

In [7]:
# Check dataframe

df_test.head()

mixed
0     a
1     b
2     1
3  True

In [8]:
# Check for mixed types data

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mixed


In [9]:
# Converting data type to string:

df_test['mixed'] = df_test['mixed'].astype('str')

In [10]:
df_test.dtypes

mixed    object
dtype: object

## Check for Missing Values

In [11]:
# Step1. Identifying missing values from products.csv

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Create a subset of the dataframe containing only the missing values in question:

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
# Check dimension of the current dataframe:

df_prods.shape

(49693, 5)

In [16]:
# Create a subset of the dataframe NOT containing the missing values:

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
# Check the dimension of the dataframe again to ensure number of rows decreased:
# New dataframe should have 16 less rows than original dataframe.

df_prods_clean.shape

(49677, 5)

#Another method to drop all missing values:

df_prods.dropna(inplace = True)



#If you wanted to use the above command to drop only the NaNs from a particular column:

df_prods.dropna(subset = [‘product_name’], inplace = True)

## Check for Duplicates

In [18]:
# Identify full duplicates within the dataframe:

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
# Displays all duplocate rows:

df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

There are 5 duplicates in products dataset.

In [20]:
# Check current number of rows in df_prods_clean dataframe:

df_prods_clean.shape

(49677, 5)

In [21]:
# Create a new dataframe that doesn’t include the duplicates:

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
# Check dimension of the dataframe does contain duplicates:

df_prods_clean_no_dups.shape

(49672, 5)

# 04. Tasks from Ex 4.5

#### 2. Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [23]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The max value in 'Department_id' column is slightly lower than expected, seems like some department_id may be missing. (Expected 13*2 = 26 as the max).

The max value in 'prices' column is much higher than the mean and the 50th percentile which needs to be investigated.

#### 3. Check for mixed-type data in your df_ords dataframe.

In [24]:
# Check for mixed types data

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There are no mixed-type data in df_ords dataframe.

#### 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

There are no mixed-type data in df_ords dataframe.

#### 5. Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [25]:
# Identify missing values from df_ords:

df_ords.isnull().sum()

Unnamed: 0                       0
order_id                         0
user_id                          0
order_numbers_users_made         0
orders_day_of_week               0
order_hour_of_day                0
days_since_prior_order      206209
dtype: int64

There are a total of 206209 missing observations in 'days_since_prior_order' column.

In [26]:
# Create a subset of the df_ords dataframe containing only the missing values:

df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [27]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_numbers_users_made  \
0                 0   2539329        1                         1   
11               11   2168274        2                         1   
26               26   1374495        3                         1   
39               39   3343014        4                         1   
45               45   2717275        5                         1   
...             ...       ...      ...                       ...   
3420930     3420930    969311   206205                         1   
3420934     3420934   3189322   206206                         1   
3421002     3421002   2166133   206207                         1   
3421019     3421019   2227043   206208                         1   
3421069     3421069   3154581   206209                         1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 7 columns]

It can observed that there are a total of 206209 missing observations in 'days_since_prior_order' column. The cause for the missing values is most likely due to no prior orders placed by those identified users above.

#### 6. Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

We should not be discarding the 'days_since_prior_order' column because those missing values hold just as much importance as the non-missing values. It tells us whether there has been prior orders place and helps us identify which order is customer's first order.

To address those missing values, we should create a new variable column containing string values "Prior_Orders" and " No_Priors_Orders" to flag which users have made prior orders.

In [28]:
# Defining a new dataframe for cleaned orders.csv data

df_ords_clean = df_ords

In [29]:
df_ords_clean['prior_orders_made'] = df_ords['days_since_prior_order'].isnull() == True

In [30]:
df_ords_clean.head()

Unnamed: 0  order_id  user_id  order_numbers_users_made  \
0           0   2539329        1                         1   
1           1   2398795        1                         2   
2           2    473747        1                         3   
3           3   2254736        1                         4   
4           4    431534        1                         5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                   2                  8                     NaN   
1                   3                  7                    15.0   
2                   3                 12                    21.0   
3                   4                  7                    29.0   
4                   4                 15                    28.0   

   prior_orders_made  
0               True  
1              False  
2              False  
3              False  
4              False

#### 7. Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [31]:
# Identify full duplicates within the dataframe:

df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [32]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_numbers_users_made, orders_day_of_week, order_hour_of_day, days_since_prior_order, prior_orders_made]
Index: []

No duplicates are found in orders dataframe.

#### 8. Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice.

In [33]:
df_ords_dups.shape

(0, 8)

There are no duplicate values. 

In [34]:
df_ords_clean.shape

(3421083, 8)

#### 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [33]:
# Exporting the cleaned product.csv

df_prods_clean_no_dups.to_csv(os.path.join(path, 'Prepared Data', 'products_checked_ex4.5.csv'))

In [34]:
# Exporting the cleaned orders.csv

df_ords_clean.to_csv(os.path.join(path,'Prepared Data', 'orders_checked_ex4.5.csv'))

#### 10. Save your Jupyter notebook and submit.